File: 프로그래머스/2/49994. 방문 길이/방문 길이.cpp
#include <string>
#include <iostream>
using namespace std;

bool isVisit[11][11][4];
int dx[4] = {0, -1, 0, 1};
int dy[4] = {-1, 0, 1, 0};

int solution(string dirs) {
    int answer = 0;
    
    int x = 5;
    int y = 5;

    
    for(int i=0; i<dirs.size(); i++){
        int dir = 0;
        char a = dirs[i];
        if(a == 'L'){ // 방향 결정
            dir = 0;
        } else if(a == 'U'){
            dir = 1;
        } else if(a == 'R'){
            dir = 2;
        } else {
            dir = 3;
        }
        
        int nx = x + dx[dir];
        int ny = y + dy[dir];
        
        if(nx < 0 || ny < 0 || nx >= 11 || ny >= 11) continue;
        int countdir = (dir + 2) % 4; 
        if(!isVisit[x][y][dir] && !isVisit[nx][ny][countdir]){
            isVisit[x][y][dir] = true;
            isVisit[nx][ny][countdir] = true;
            answer++;
        }
        x = nx;
        y = ny;
    }
    
    return answer;
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.*;

class Solution {
    
    int[] dx = {-1, 0, 1, 0};
    int[] dy = {0, -1, 0, 1};
    boolean[][][] isVisit = new boolean[11][11][4];
    
    public int solution(String dirs) {
        int answer = 0;
        
        int x = 5;
        int y = 5;
        
        for(int i=0; i<dirs.length(); i++){
            int dir = 0;
            if(dirs.charAt(i) == 'U'){
                dir = 0;
            }
            else if(dirs.charAt(i) == 'D'){
                dir = 2;
            }
            else if(dirs.charAt(i) == 'L'){
                dir = 1;
            }
            else{
                dir = 3;
            }
            
            int nx = x + dx[dir];
            int ny = y + dy[dir];
            
            if(nx >= 0 && ny >= 0 && nx <= 10 && ny <= 10){
                int cntDir = (dir+2) % 4;
                if(!isVisit[y][x][dir] && !isVisit[ny][nx][cntDir]){
                    isVisit[y][x][dir] = true;
                    isVisit[ny][nx][cntDir] = true;
                    answer++;
                }
                x = nx;
                y = ny;
            }
        }
        
        
        return answer;
    }
}